copy-conditions.go

```diff
@@ -91,9 +75,37 @@ func (c *CopyConditions) SetModified(modTime time.Time) error {
 	if modTime.IsZero() {
 		return ErrInvalidArgument("Modified since cannot be empty.")
 	}
-	c.conditions = append(c.conditions, copyCondition{
-		key:   "x-amz-copy-source-if-modified-since",
-		value: modTime.Format(http.TimeFormat),
-	})
+	c.conditions["x-amz-copy-source-if-modified-since"] = modTime.Format(http.TimeFormat)
 	return nil
 }
+
+// SetByteRange - set the start and end of the source object to be
+// copied.
+func (c *CopyConditions) SetByteRange(start, end int64) error {
+	if start < 0 || end < start {
+		return ErrInvalidArgument("Range start less than 0 or range end less than range start.")
+	}
+	if end-start+1 < 1 {
+		return ErrInvalidArgument("Offset must refer to a non-zero range length.")
+	}
+	c.byteRangeEnd = end
+	c.byteRangeStart = start
+	return nil
+}
+
+func (c *CopyConditions) getRangeSize() int64 {
+	if c.byteRangeStart < 0 {
+		// only happens if byte-range was not set by user
+		return 0
```

Comment:
Probably it is better to set this to -1 to say unknown. Probably zero is a valid value.